### PR TITLE
Adding `asuint`  implementation to hlsl

### LINKED ISSUE #70097: Implement the `asuint` HLSL Function
- [ ] Implement `asuint` clang builtin,
- [ ] Link `asuint` clang builtin with `hlsl_intrinsics.h`
- [ ] Add sema checks for `asuint` to `CheckHLSLBuiltinFunctionCall` in `SemaChecking.cpp`
- [ ] Add codegen for `asuint` to `EmitHLSLBuiltinExpr` in `CGBuiltin.cpp`
- [ ] Add codegen tests to `clang/test/CodeGenHLSL/builtins/asuint.hlsl`
- [ ] Add sema tests to `clang/test/SemaHLSL/BuiltIns/asuint-errors.hlsl`

## DirectX

There were no DXIL opcodes found for  most `asuint` cases. There is an opcode for the three arg case:
| DXIL Opcode | DXIL OpName | Shader Model | Shader Stages | Description |
| ----------- | ----------- | ------------ | ------------- |------------- |
| 102 | SplitDouble| 6.0 | () | splits a double into low and high parts

## SPIR-V

There is no support for `asuint` when targeting SPIR-V.

### Example 1
```hlsl
//dxc asuint_test.hlsl -T lib_6_8 -enable-16bit-types -O0

export uint4 fn(double4 p1, uint4 p2, uint4 p3) {
    asuint(p1, p2, p3);
    return p3;
}
```

 ### Example 2
```hlsl
//dxc asuint_1_test.hlsl -T lib_6_8 -enable-16bit-types -O0

export uint4 fn(float p1) {
    return asuint(p1);
}
```
 ### Example 3
```hlsl
//dxc asuint_2_test.hlsl -T lib_6_8 -enable-16bit-types -O0

export uint4 fn(uint p1) {
    return asuint(p1);
}
```
 ### Example 4
```hlsl
//dxc asuint_3_test.hlsl -T lib_6_8 -enable-16bit-types -O0

export uint4 fn(int p1) {
    return asuint(p1);
}
```
## HLSL:

Interprets the bit pattern of *x* as an unsigned integer.

| ret asuint(*x*) |
|-----------------|

 

## Parameters

| Item                                                   | Description                        |
|--------------------------------------------------------|------------------------------------|
| <span id="x"></span><span id="X"></span>*x*<br/> | \[in\] The input value.<br/> |

 

## Return Value

The input interpreted as an unsigned integer.

## Type Description

| Name  | [**Template Type**](https://github.com/MicrosoftDocs/win32/blob/docs/desktop-src//direct3dhlsl/dx-graphics-hlsl-intrinsic-functions.md)                                                  | [**Component Type**](https://github.com/MicrosoftDocs/win32/blob/docs/desktop-src//direct3dhlsl/dx-graphics-hlsl-intrinsic-functions.md)                 | Size                           |
|-------|----------------------------------------------------------------------------------------------------------------|--------------------------------------------------------------------------------|--------------------------------|
| *x*   | [**scalar**](https://github.com/MicrosoftDocs/win32/blob/docs/desktop-src//direct3dhlsl/dx-graphics-hlsl-intrinsic-functions.md), **vector**, or **matrix** | [**float**](/windows/desktop/WinProg/windows-data-types), [**int**](/windows/desktop/WinProg/windows-data-types) | any                            |
| *ret* | same as input *x*                                                                                              | [**uint**](/windows/desktop/WinProg/windows-data-types)                                         | same dimension(s) as input *x* |

 

## Minimum Shader Model

This function is supported in the following shader models.

| Shader Model                                                        | Supported |
|---------------------------------------------------------------------|-----------|
| [Shader Model 4](https://github.com/MicrosoftDocs/win32/blob/docs/desktop-src//direct3dhlsl/dx-graphics-hlsl-sm4.md) and higher shader models | yes       |
| [Shader Model 3 (DirectX HLSL)](https://github.com/MicrosoftDocs/win32/blob/docs/desktop-src//direct3dhlsl/dx-graphics-hlsl-sm3.md)           | no        |
| [Shader Model 2 (DirectX HLSL)](https://github.com/MicrosoftDocs/win32/blob/docs/desktop-src//direct3dhlsl/dx-graphics-hlsl-sm2.md)           | no        |
| [Shader Model 1 (DirectX HLSL)](https://github.com/MicrosoftDocs/win32/blob/docs/desktop-src//direct3dhlsl/dx-graphics-hlsl-sm1.md)           | no        |

 

## See also

<dl> <dt>

[**Intrinsic Functions (DirectX HLSL)**](https://github.com/MicrosoftDocs/win32/blob/docs/desktop-src//direct3dhlsl/dx-graphics-hlsl-intrinsic-functions.md)
</dt> </dl>

### PR DESCRIPTION
**Pull request by joaosaffran**
Implements support for the `asuint` HLSL function casting behaviour.
Addressing the `splitdouble` scenario will be addressed in a future PR. 

Fixes: #70097

### CODE REVIEW

File: clang/lib/Headers/hlsl/hlsl_intrinsics.h [hunk 1]
```diff
@@ -387,6 +387,23 @@ float3 asin(float3);
...
+
+_HLSL_BUILTIN_ALIAS(__builtin_hlsl_elementwise_asuint)
+uint asuint(float);
+_HLSL_BUILTIN_ALIAS(__builtin_hlsl_elementwise_asuint)
+uint2 asuint(float2);
+_HLSL_BUILTIN_ALIAS(__builtin_hlsl_elementwise_asuint)
+uint3 asuint(float3);
+_HLSL_BUILTIN_ALIAS(__builtin_hlsl_elementwise_asuint)
+uint4 asuint(float4);
+
```
Comment (farzonl) [inline clang/lib/Headers/hlsl/hlsl_intrinsics.h:406]: Need to handle the int case. and i'm thinking for the uint case maybe just return the input? we haven't done that in the header before, but the uint case doesn't warrant code gen.

Comment (bogner) [reply clang/lib/Headers/hlsl/hlsl_intrinsics.h:406]: `IRBuilder::CreateBitCast` just returns the input if the type doesn't change, so the logic may already to the right thing.

Comment (farzonl) [reply clang/lib/Headers/hlsl/hlsl_intrinsics.h:406]: In that case add the int and uint cases in the header and add the equivalent tests to asuint.hlsl

File: clang/lib/Sema/SemaHLSL.cpp [hunk 2]
```diff
@@ -1427,6 +1447,14 @@ bool CheckAllArgsHaveFloatRepresentation(Sema *S, CallExpr *TheCall) {
                                   checkAllFloatTypes);
 }
 
+bool CheckArgIsFloatOrIntWithoutImplicits(Sema *S, Expr *Arg) {
+  auto checkFloat = [](clang::QualType PassedType) -> bool {
```
Comment (farzonl) [inline clang/lib/Sema/SemaHLSL.cpp:1451]: LLVM style guide is this should be capitalized. Also wouldn't it make more sense as `CheckFloatOrInt`

File: clang/lib/Sema/SemaHLSL.cpp [hunk 3]
```diff
@@ -1427,6 +1447,14 @@ bool CheckAllArgsHaveFloatRepresentation(Sema *S, CallExpr *TheCall) {
...
+    return !PassedType->isFloat32Type() && !PassedType->isIntegerType();
```
Comment (farzonl) [inline clang/lib/Sema/SemaHLSL.cpp:1452]:
```suggestion 
    return !(PassedType->isFloat32Type() || PassedType->isIntegerType());
```

File: clang/test/CodeGenHLSL/builtins/asuint.hlsl [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+// CHECK: %1 = bitcast float %0 to i32
+export uint test_asuint_float(float p0) {
+  return asuint(p0);
+}
+
+// CHECK-LABEL: test_asuint_float
+// CHECK: %1 = bitcast <4 x float> %0 to <4 x i32>
+export uint4 test_asuint_float4(float4 p0) {
+  return asuint(p0);
+}
```
Comment (farzonl) [inline clang/test/CodeGenHLSL/builtins/asuint.hlsl:26]: add a new line

**Review by farzonl — APPROVED** [review]
LGTM

File: clang/test/CodeGenHLSL/builtins/asuint.hlsl [hunk 5]
```diff
@@ -0,0 +1,26 @@
+// RUN: %clang_cc1 -finclude-default-header -x hlsl -triple dxil-pc-shadermodel6.3-library %s -fnative-half-type -emit-llvm -disable-llvm-passes -o - | FileCheck %s
+
+
+// CHECK-LABEL: test_asuint4_uint
+// CHECK: ret i32 %0
+export uint test_asuint4_uint(uint p0) {
+  return asuint(p0);
+}
```
Comment (bogner) [inline clang/test/CodeGenHLSL/builtins/asuint.hlsl:8]: This doesn't really test the function effectively, it just checks that we return an `i32`. We can use a couple of tools in FileCheck to do better, notably [capturing variables](https://llvm.org/docs/CommandGuide/FileCheck.html#filecheck-string-substitution-blocks) and [CHECK-NOT](https://llvm.org/docs/CommandGuide/FileCheck.html#the-check-not-directive).

Now, we want to make sure that the input of the function is returned, and that we don't see a bitcast in between.

Take a look at the output of the test command by running it manually. Assuming we update the flags as I suggested above, we see something like this:
```llvm
define noundef i32 @"?test_uint@@YAII@Z"(i32 noundef returned %p0) local_unnamed_addr #0 {
entry:
  ret i32 %p0
}
```

Getting the name of the input variable is a little bit awkward because (1) we mangle the function name, and (2) we get some attributes applied that we don't really want the test to be sensitive to, so need a somewhat loose regex. Then we want to check that no bitcast occurs, and that we return the input directly:

```c++
// CHECK: define {{.*}}test_asuint4_uint{{.*}}(i32{{.*}} [[VAL:%.*]])
// CHECK-NOT: bitcast
// CHECK: ret i32 [[VAL]]
```

File: clang/lib/Headers/hlsl/hlsl_intrinsics.h [hunk 6]
```diff
@@ -387,6 +387,23 @@ float3 asin(float3);
...
+/// \param Val The input value.
+
+_HLSL_BUILTIN_ALIAS(__builtin_hlsl_elementwise_asuint)
+uint asuint(float);
+_HLSL_BUILTIN_ALIAS(__builtin_hlsl_elementwise_asuint)
+uint2 asuint(float2);
+_HLSL_BUILTIN_ALIAS(__builtin_hlsl_elementwise_asuint)
+uint3 asuint(float3);
+_HLSL_BUILTIN_ALIAS(__builtin_hlsl_elementwise_asuint)
+uint4 asuint(float4);
```
Comment (bogner) [inline clang/lib/Headers/hlsl/hlsl_intrinsics.h:405]: Shouldn't we have definitions for `uint asuint(int)`, `uint asuint(uint)`, etc?

File: clang/test/CodeGenHLSL/builtins/asuint.hlsl [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+
+// CHECK-LABEL: test_asuint4_int
+// CHECK: %splat.splatinsert = insertelement <4 x i32> poison, i32 %0, i64 0
+export uint4 test_asuint4_int(int p0) {
```
Comment (bogner) [inline clang/test/CodeGenHLSL/builtins/asuint.hlsl:12]: This test takes an int input but returns a uint4 - I don't think it's doing what you intend. There is an implicit cast *after* the asuint call to convert to a vector of 4 elements.

**Review by bogner — COMMENTED** [review]
This is going in the right direction but the tests need some work. I made a couple of comments to point you in the right direction.

We also need to make sure we're testing all of the cases we care about, so we need tests with inputs of `int`, `uint`, vector of `int`, vector of `uint`, `float`, vector of `float`. You've covered some but not all of these.

**Review by llvm-beanz — COMMENTED** [review]
One question I have about this is do we actually need new builtins? These are just doing bit casts correct?

Could we implement them entirely in the header? Something like:

```hlsl
template <typename T, typename U>
T bit_cast(U Val) { return __builtin_bit_cast(T, Val); }

uint asuint(float F) { return bit_cast<uint>(F); }
```

This seems to work for me in godbolt: https://godbolt.org/z/8b61a1WMr

We may want to apply the `[[alwaysinline]]` attribute, but it seems like a reasonable approach. Thoughts?

File: clang/lib/Headers/hlsl/hlsl_intrinsics.h (hunk 6 above)
Comment (joaosaffran) [reply clang/lib/Headers/hlsl/hlsl_intrinsics.h:405]: When I add the `uint asuint(int)` and `uint asuint(uint)`, I am seeing the following error:

```
/workspace/llvm-project/clang/test/SemaHLSL/BuiltIns/asuint-errors.hlsl:11:12: error: call to 'asuint' is ambiguous
   11 |     return asuint(p1);
      |            ^~~~~~
/workspace/llvm-project/build-debug/lib/clang/20/include/hlsl/hlsl_intrinsics.h:399:6: note: candidate function
  399 | uint asuint(float);
      |      ^
/workspace/llvm-project/build-debug/lib/clang/20/include/hlsl/hlsl_intrinsics.h:408:6: note: candidate function
  408 | uint asuint(uint);
      |      ^
/workspace/llvm-project/build-debug/lib/clang/20/include/hlsl/hlsl_intrinsics.h:417:6: note: candidate function
  417 | uint asuint(int);
```

It seems like clang is not being able to figure out which function to call. Please let me know if you have a solution for this issue.

Comment (bogner) [reply clang/lib/Headers/hlsl/hlsl_intrinsics.h:405]: This is just overload resolution rules. Since there are two equally bad options to convert to from a `double` (`float` and `int` both lose information) it doesn't know which one to call and (correctly) errors out.

Comment (farzonl) [reply clang/lib/Headers/hlsl/hlsl_intrinsics.h:405]: > Shouldn't we have definitions for `uint asuint(int)`, `uint asuint(uint)`, etc?

I thought i mentioned that. but my conversation got resolved

Comment (joaosaffran) [reply clang/lib/Headers/hlsl/hlsl_intrinsics.h:405]: I added, the requested definition, however I had to change table gen definitions for this builtin

File: clang/test/SemaHLSL/BuiltIns/asuint-errors.hlsl [hunk 8]
```diff
@@ -0,0 +1,18 @@
+// RUN: %clang_cc1 -finclude-default-header -triple dxil-pc-shadermodel6.6-library %s -fnative-half-type -emit-llvm-only -disable-llvm-passes -verify -verify-ignore-unexpected
```
Comment (bogner) [inline clang/test/SemaHLSL/BuiltIns/asuint-errors.hlsl:1]: I don't think we should be using `-verify-ignore-unexpected` here - it makes it harder to understand what's going wrong if the test fails, and generally we shouldn't have errors that we're not checking for in a test.

File: clang/test/SemaHLSL/BuiltIns/asuint-errors.hlsl [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+export uint4 test_asuint_too_many_arg(float p0, float p1) {
+  return __builtin_hlsl_elementwise_asuint(p0, p1);
+  // expected-error@-1 {{too many arguments to function call, expected 1, have 2}}
+}
+
+
+export uint fn(double p1) {
+    return asuint(p1);
+    // expected-error@-1 {{passing 'double' to parameter of incompatible type 'float'}}
+}
```
Comment (bogner) [inline clang/test/SemaHLSL/BuiltIns/asuint-errors.hlsl:13]: I suspect this test is failing because we won't actually get to the error handling on the builtin - since `asuint` has `float` and `int` overloads, this call should be ambiguous.

**Comment by joaosaffran** [comment]
@llvm-beanz I've implemented your suggestions. One issue I found is that the code generation will work with half types, which are not supported in the current implementation. Any suggestions into how to approach this?

**Comment by farzonl** [comment]
> @llvm-beanz I've implemented your suggestions. One issue I found is that the code generation will work with half types, which are not supported in the current implementation. Any suggestions into how to approach this?

is there a way to know if the use of `__builtin_bit_cast` came fro `asuint`? like maybe by checking if the return type is already set to uint?

if so then you could disallow  `half` types in HLSLSema.

File: clang/test/CodeGenHLSL/builtins/asuint.hlsl [hunk 10]
```diff
@@ -0,0 +1,41 @@
+// RUN: %clang_cc1 -finclude-default-header -x hlsl -triple dxil-pc-shadermodel6.3-library %s -fnative-half-type -emit-llvm -O1 -o - | FileCheck %s
+
+// CHECK: define {{.*}}test_uint{{.*}}(i32 {{.*}} [[VAL:%.*]]){{.*}} 
+// CHECK-NOT: bitcast
+// CHECK: ret i32 [[VAL]]
+export uint test_uint(uint p0) {
```
Comment (farzonl) [inline clang/test/CodeGenHLSL/builtins/asuint.hlsl:6]: We haven't been putting export on the other intrinsic test cases. The issue had it because it wouldn't run without it in DXC.

File: clang/lib/Headers/hlsl/hlsl_intrinsics.h [hunk 11]
```diff
@@ -387,6 +387,65 @@ float3 asin(float3);
...
 
+//===----------------------------------------------------------------------===//
+// asuint builtins
+//===----------------------------------------------------------------------===//
+
+/// \fn uint asuint(T Val)
+/// \brief Interprets the bit pattern of x as an unsigned integer.
+/// \param Val The input value.
+
+__attribute__((__always_inline__, __nodebug__)) static inline uint
```
Comment (farzonl) [inline clang/lib/Headers/hlsl/hlsl_intrinsics.h:398]: I don't really like how dissimilar this looks to all the other intrinsics. Maybe we could capture these new attributes in a new preprocessor macro. Something like `_HLSL_CAST_BUILTIN_ALIAS`?

**Comment by joaosaffran** [comment]
@farzonl I discussed the `half` type support with @llvm-beanz, he suggested using static asserts, which I've added and now checks whether the type being passed has 32bits. This prevents the `half` support. Any concerns regarding this approach?

File: clang/lib/Headers/hlsl/hlsl_intrinsics.h (hunk 11 above)
Comment (joaosaffran) [reply clang/lib/Headers/hlsl/hlsl_intrinsics.h]: @farzonl Let me know if you still think this is necessary after changing the implementation to use templates.

Comment (bogner) [reply clang/lib/Headers/hlsl/hlsl_intrinsics.h]: I kind of disagree with this, and I find the version with the macro pretty hard to follow.
1. Even if we have a macro to help with the attributes or something, it shouldn't be called "alias" - this isn't an alias at all (because of the static assert).
2. I don't really like that this single macro is defining multiple functions. The uses of `_HLSL_BUILTIN_ALIAS` decorate the definitions but each definition is written out clearly in source.

I could see some value in putting the `__attribute__((__always_inline__, __nodebug__))` bit in a macro so that it's easy to be consistent, but I don't think the function definitions themselves should be done this way.

Comment (farzonl) [reply clang/lib/Headers/hlsl/hlsl_intrinsics.h]: Justin is right, the static assert makes this no longer just an alias for `__builtin_bit_cast` so the macro name has to change at a minimum. 

Also we mostly used macros to cover our attribute usage. So this is as much a new pattern as what you had. That said, while this isn't what I imagined this to look like I'm not opposed to how this turned out if the function body you defined is going to be reused across all our cast apis? That all depends on if the remaining cast work is boilerplate enough that this is flexible enough to support it.

Comment (joaosaffran) [reply clang/lib/Headers/hlsl/hlsl_intrinsics.h]: I think there might be some things we are unaware of right now. I think we can use the macros to host the attributes, or help us do some of the `static_cast`.  In the future, once we've implemented more of the casting functionality, we can reassess and see if there is more boilerplate that can be moved into the macro.

**Comment by joaosaffran** [comment]
After discussing with @bogner and @farzonl, we decided the following: 

- Using `static_assert` for the errors makes the user experience inconsistent, since we expect the type checking to happen in `SemaHLSL.cpp`, as other builtins.
- The 16 bit casts don't have this problem (because that would be a lossy conversion).
- Therefore, we should introduce a builtin for hlsl-specific 32-bit casts, this can be reused to implement `asfloat`.

File: clang/test/SemaHLSL/BuiltIns/asuint-errors.hlsl [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+
+
+uint4 test_asuint_too_many_arg(float p0, float p1) {
+  return asuint(p0, p1);
+  // expected-error@-1 {{no matching function for call to 'asuint'}}
+}
+
+uint test_asuint_double(double p1) {
+    return asuint(p1);
+    // expected-error@-1{clang/20/include/hlsl/hlsl_details.h} {{static assertion failed due to requirement 'sizeof(unsigned int) == sizeof(double)'}}
```
Comment (farzonl) [inline clang/test/SemaHLSL/BuiltIns/asuint-errors.hlsl:11]: This is going to fail in a year when the version goes to 21. the error check should stick to the string in the static assert.
```suggestion
    // expected-error@-1{{.*}} {{static assertion failed due to requirement 'sizeof(unsigned int) == sizeof(double)'}}
```

File: clang/lib/Headers/hlsl/hlsl_details.h [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+struct enable_if<true, T> {
+  using Type = T;
+};
+
+template<bool B, typename T=void>
+using enable_if_t = typename enable_if<B,T>::Type;
+
+template<typename U, typename T, int N>
+HLSL_BUILTIN_ATTRIBUTES vector<U, N> bit_cast(vector<T, N> V) {
+  _Static_assert(sizeof(U) == sizeof(T), "casting types must be same bit size.");
```
Comment (llvm-beanz) [inline clang/lib/Headers/hlsl/hlsl_details.h:30]: Try this. The 2021 language mode is still based on C++98, so enable_if_t is finacky. We can clean it up in a subsequent PR once I finish fixing the language modes.

```suggestion
HLSL_BUILTIN_ATTRIBUTES
typename enable_if<sizeof(U) == sizeof(T), vector<U, N> >::Type
bit_cast(vector<T, N> V) {
```

File: clang/lib/Headers/hlsl/hlsl_intrinsics.h [hunk 14]
```diff
@@ -10,6 +10,7 @@
 #define _HLSL_HLSL_INTRINSICS_H_
 
 namespace hlsl {
+#include "hlsl_details.h"
```
Comment (bogner) [inline clang/lib/Headers/hlsl/hlsl_intrinsics.h:13]: We shouldn't `#include` a file inside a namespace. Instead, "hlsl_details.h" should also have a `namespace hlsl {}` block around the `__detail` namespace.

File: clang/lib/Headers/hlsl/hlsl_details.h [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _HLSL_HLSL_DETAILS_H_
+#define _HLSL_HLSL_DETAILS_H_
+
+namespace __details {
```
Comment (bogner) [inline clang/lib/Headers/hlsl/hlsl_details.h:12]: If we're trying to mimic libc++ style this would be `__detail` rather than `__details`

File: clang/lib/Headers/hlsl/hlsl_details.h [hunk 16]
```diff
@@ -0,0 +1,35 @@
+//===----- hlsl_intrinsics.h - HLSL definitions for intrinsics ----------===//
```
Comment (bogner) [inline clang/lib/Headers/hlsl/hlsl_details.h:1]: This comment needs updating.

File: clang/lib/Headers/hlsl/hlsl_details.h [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _HLSL_HLSL_DETAILS_H_
+#define _HLSL_HLSL_DETAILS_H_
+
+namespace __details {
+#define HLSL_INLINE_ATTRIBUTE                                                  \
+  __attribute__((__always_inline__, __nodebug__)) static inline
```
Comment (bogner) [inline clang/lib/Headers/hlsl/hlsl_details.h:14]: This should start with an underscore to avoid collisions with user's defines (`_CapitalLetter` is reserved for implementations). Since it isn't just an attribute (it's also `static inline` I'd just call it `_HLSL_INLINE`.

File: clang/test/SemaHLSL/BuiltIns/asuint-errors.hlsl (hunk 8 above)
Comment (bogner) [reply clang/test/SemaHLSL/BuiltIns/asuint-errors.hlsl:1]: This still has `-verify-ignore-unexpected`...

Comment (joaosaffran) [reply clang/test/SemaHLSL/BuiltIns/asuint-errors.hlsl:1]: I am seeing some issues when testing without it, for example:

```
error: 'expected-error' diagnostics seen but not expected
```

The error messages are really verbose. What is the best approach here: i) add the verbose lines?; ii) ignore the unexpected and check only one of the errors?

Comment (llvm-beanz) [reply clang/test/SemaHLSL/BuiltIns/asuint-errors.hlsl:1]: what messages are seen but not expected?

Comment (joaosaffran) [reply clang/test/SemaHLSL/BuiltIns/asuint-errors.hlsl]: I've updated the tests, but here is the full error message, for such tests:

```hlsl
/workspace/llvm-project/build-debug/bin/clang -cc1 -internal-isystem /workspace/llvm-project/build-debug/lib/clang/20/include -nostdsysteminc -finclude-default-header -x hlsl -triple dxil-pc-shadermodel6.3-library /workspace/llvm-project/clang/test/SemaHLSL/BuiltIns/asuint-errors.hlsl -fnative-half-type -emit-llvm 
[12/12] Creating executable symlink bin/clang
/workspace/llvm-project/clang/test/SemaHLSL/BuiltIns/asuint-errors.hlsl:5:10: error: no matching function for call to 'asuint'
    5 |   return asuint(p0, p1);
      |          ^~~~~~
/workspace/llvm-project/build-debug/lib/clang/20/include/hlsl/hlsl_intrinsics.h:401:30: note: candidate function template not viable: requires single argument 'V', but 2 arguments were provided
  401 | _HLSL_INLINE vector<uint, N> asuint(vector<T, N> V) {
      |                              ^      ~~~~~~~~~~~~~~
/workspace/llvm-project/build-debug/lib/clang/20/include/hlsl/hlsl_intrinsics.h:405:41: note: candidate function template not viable: requires single argument 'F', but 2 arguments were provided
  405 | template <typename T> _HLSL_INLINE uint asuint(T F) {
      |                                         ^      ~~~
/workspace/llvm-project/build-debug/lib/clang/20/include/hlsl/hlsl_intrinsics.h:406:10: error: no matching function for call to 'bit_cast'
  406 |   return __detail::bit_cast<uint, T>(F);
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~
/workspace/llvm-project/clang/test/SemaHLSL/BuiltIns/asuint-errors.hlsl:12:12: note: in instantiation of function template specialization 'hlsl::asuint<double>' requested here
   12 |     return asuint(p1);
      |            ^
/workspace/llvm-project/build-debug/lib/clang/20/include/hlsl/hlsl_detail.h:27:1: note: candidate template ignored: could not match 'vector<double, N>' against 'double'
   27 | bit_cast(vector<T, N> V) {
      | ^
/workspace/llvm-project/build-debug/lib/clang/20/include/hlsl/hlsl_detail.h:32:66: note: candidate template ignored: substitution failure [with U = uint, T = double]: no type named 'Type' in 'hlsl::__detail::enable_if<false, unsigned int>'
   32 | _HLSL_INLINE typename enable_if<sizeof(U) == sizeof(T), U>::Type bit_cast(T F) {
      |              ~~~~~~~~                                            ^
In file included from <built-in>:1:
In file included from /workspace/llvm-project/build-debug/lib/clang/20/include/hlsl.h:13:
/workspace/llvm-project/build-debug/lib/clang/20/include/hlsl/hlsl_intrinsics.h:406:10: error: no matching function for call to 'bit_cast'
  406 |   return __detail::bit_cast<uint, T>(F);
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~
/workspace/llvm-project/clang/test/SemaHLSL/BuiltIns/asuint-errors.hlsl:20:12: note: in instantiation of function template specialization 'hlsl::asuint<half>' requested here
   20 |     return asuint(p1);
      |            ^
/workspace/llvm-project/build-debug/lib/clang/20/include/hlsl/hlsl_detail.h:27:1: note: candidate template ignored: could not match 'vector<half, N>' against 'half'
   27 | bit_cast(vector<T, N> V) {
      | ^
/workspace/llvm-project/build-debug/lib/clang/20/include/hlsl/hlsl_detail.h:32:66: note: candidate template ignored: substitution failure [with U = uint, T = half]: no type named 'Type' in 'hlsl::__detail::enable_if<false, unsigned int>'
   32 | _HLSL_INLINE typename enable_if<sizeof(U) == sizeof(T), U>::Type bit_cast(T F) {
      |              ~~~~~~~~                                            ^
3 errors generated.
```

**Comment by joaosaffran** [comment]
> ⚠️ C/C++ code formatter, clang-format found issues in your code. ⚠️
> 
> You can test this locally with the following command:
> View the diff from clang-format here.

This is formatting issue is caused because, clang's LLVM style assumes C++14, where the space isn't required, but HLSL is C++98 where the space is required. This issue will be addressed once we address all fixes for future HLSL releases.

File: clang/test/SemaHLSL/BuiltIns/asuint-errors.hlsl [hunk 18]
```diff
@@ -0,0 +1,25 @@
+// RUN: %clang_cc1 -finclude-default-header -triple dxil-pc-shadermodel6.6-library %s -fnative-half-type -emit-llvm-only -disable-llvm-passes -verify
```
Comment (bogner) [inline clang/test/SemaHLSL/BuiltIns/asuint-errors.hlsl:1]: The `-emit-llvm-only` and `-disable-llvm-passes` flags don't really do anything when combined with verify. We don't need them.
```suggestion
// RUN: %clang_cc1 -finclude-default-header -triple dxil-pc-shadermodel6.6-library %s -fnative-half-type -verify
```

**Review by bogner — APPROVED** [review]
looks good, thanks!

**Review by farzonl — APPROVED** [review]
